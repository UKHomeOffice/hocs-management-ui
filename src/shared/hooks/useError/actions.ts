import { FormError } from '../../models/formError';

export type AddFormError = {
    payload: FormError;
    type: 'AddFormError';
};

export type ClearError = {
    type: 'ClearError'
};

export type SetTitle = {
    payload: string;
    type: 'SetTitle';
};

export type SetDescription = {
    payload: string;
    type: 'SetDescription';
};

export type Action =
    AddFormError |
    ClearError |
    SetDescription |
    SetTitle;
