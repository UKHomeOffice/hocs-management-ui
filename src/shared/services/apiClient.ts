import axios from 'axios';

export default {
    createClient: () => {
        const client = axios.create();

        return {
            get: client.get,
            delete: client.delete,
            post: client.post,
            put: client.put
        };
    }
};
