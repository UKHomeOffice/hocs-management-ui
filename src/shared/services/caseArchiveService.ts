import axios from 'axios';
import ArchiveCaseModel from '../models/archiveCaseModel';

export const archiveCase = (archiveCaseModel: ArchiveCaseModel) => new Promise((resolve, reject) => axios
    .post('/api/case/archive', archiveCaseModel)
    .then(() => resolve(null))
    .catch(reason => reject(reason))
);
