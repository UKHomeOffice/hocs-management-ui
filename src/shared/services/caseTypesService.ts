import axios from 'axios';
import CaseType from '../models/caseType';
import Item from '../models/item';

const sortByLabel: ((a: Item, b: Item) => number) = (a, b) => {
    const labelA = a.label.toLocaleLowerCase();
    const labelB = b.label.toLocaleLowerCase();
    let comparison = 0;
    if (labelA > labelB) {
        comparison = 1;
    } else if (labelA < labelB) {
        comparison = -1;
    }
    return comparison;
};

export const getCaseTypes = () => new Promise<CaseType[]>((resolve, reject) =>
    axios.get('/api/case-types')
        .then((response) => {
            resolve(response.data.sort(sortByLabel));
        })
        .catch(error => reject(error))
);
