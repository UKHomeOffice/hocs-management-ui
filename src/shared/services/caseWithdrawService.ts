import axios from 'axios';
import WithdrawCaseModel from '../models/withdrawCaseModel';

export const withdrawCase = (withdrawCaseModel: WithdrawCaseModel) => new Promise((resolve, reject) => axios
    .post('/api/case/withdraw', withdrawCaseModel)
    .then(() => resolve(null))
    .catch(reason => reject(reason))
);
