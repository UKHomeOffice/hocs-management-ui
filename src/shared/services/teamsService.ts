import axios from 'axios';
import Item from '../models/item';
import Team from '../models/team';

export const getTeam = (teamId: string) => new Promise<Team>((resolve, reject) =>
    axios.get(`/api/teams/${teamId}`)
        .then(response => resolve(response.data))
        .catch(error => reject(error))
);

export const getTeams = () => new Promise<[Item]>((resolve, reject) =>
    axios.get('/api/teams')
        .then(response => resolve(response.data))
        .catch(error => reject(error))
);

export const getTeamMembers = (teamId: string) => new Promise((resolve, reject) => axios
    .get(`/api/teams/${teamId}/members`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const getTeamsForUser = (userId: string) => new Promise<[Item]>((resolve, reject) => axios
    .get(`/api/teams/${userId}/teams`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const createTeam = (team: Team) => new Promise((resolve, reject) => axios
    .post(`/api/teams/unit/${team.unitUUID}`, team)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);
