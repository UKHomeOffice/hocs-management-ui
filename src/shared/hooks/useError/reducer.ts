import { Action } from './actions';
import PageError from '../../models/pageError';

export const reducer = (state: PageError, action: Action): PageError => {
    switch (action.type) {
        case 'AddFormError':
            return {
                error: { ...state.error, formErrors: [...state.error ? state.error.formErrors || [] : [], action.payload] }
            };
        case 'ClearError':
            return { error: undefined };
        case 'SetDescription':
            return { error: { ...state.error, description: action.payload } };
        case 'SetTitle':
            return { error: { ...state.error, title: action.payload } };
    }
    return state;
};
