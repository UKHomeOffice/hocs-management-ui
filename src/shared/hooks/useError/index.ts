import React, { Reducer } from 'react';
import PageError from '../../models/pageError';
import { Action } from './actions';
import { reducer } from './reducer';
import { FormError } from '../../models/formError';
import ErrorMessage from 'shared/models/errorMessage';

const useError = (formErrorDescription?: string, formErrorTitle?: string)
    : [
        PageError,
        (formError: FormError) => void,
        () => void,
        (error: ErrorMessage) => void
    ] => {
    const [pageError, dispatch] = React.useReducer<Reducer<PageError, Action>>(reducer, {});

    const addFormError = (formError: FormError) => {
        (formErrorDescription || formErrorTitle) && setMessage({ title: formErrorTitle || '', description: formErrorDescription || '' });
        dispatch({ type: 'AddFormError', payload: formError });
    };
    const clearErrors = () => dispatch({ type: 'ClearError' });
    const setMessage = (errorMessage: ErrorMessage) => {
        dispatch({ type: 'SetDescription', payload: errorMessage.description });
        dispatch({ type: 'SetTitle', payload: errorMessage.title });
    };
    return [pageError, addFormError, clearErrors, setMessage];
};

export default useError;
