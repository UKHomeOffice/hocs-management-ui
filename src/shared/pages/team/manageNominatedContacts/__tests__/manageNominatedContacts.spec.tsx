import React from 'react';

import * as useError from '../../../../hooks/useError';
import { match, MemoryRouter } from 'react-router';
import ManageNominatedContacts from '../manageNominatedContacts';
import { render, RenderResult, waitFor } from '@testing-library/react';
import { createBrowserHistory, History, Location } from 'history';
import Team from '../../../../models/team';
import { act } from 'react-dom/test-utils';
import * as TeamsService from '../../../../services/teamsService';

const useErrorSpy = jest.spyOn(useError, 'default');

let match: match<any>;
let history: History<any>;
let location: Location;
let mockState: Team;

const useStateSpy = jest.spyOn(React, 'useState');
const mockSetState = jest.fn();
const getTeamSpy = jest.spyOn(TeamsService, 'getTeam');
const pageErrorSpy = jest.fn();
const setErrorMessageSpy = jest.fn();

const renderComponent = () => render(
    <MemoryRouter>
        <ManageNominatedContacts history={history} location={location} match={match}></ManageNominatedContacts>
    </MemoryRouter>
);

jest.mock('../nominatedContactList', () => () => 'MOCK_CONTACT_LIST_COMPONENT');
jest.mock('../addNominatedContacts', () => () => 'MOCK_ADD_NOMINATED_CONTACT_COMPONENT');


describe('manage nominated contacts', () => {
    beforeEach(() => {
        history = createBrowserHistory();

        match = {
            isExact: true,
            params: { teamId: '__teamId__' },
            path: '',
            url: ''
        };

        location = {
            hash: '',
            key: '',
            pathname: '',
            search: '',
            state: {}
        };

        mockState = {
            active: true,
            displayName: '__teamName__',
            letterName: '',
            permissions: [],
            type: ''
        };

        useStateSpy.mockImplementation(() => [mockState, mockSetState]);

        useErrorSpy.mockImplementation(() => [{}, jest.fn(), jest.fn(), setErrorMessageSpy]);
        pageErrorSpy.mockReset();
        setErrorMessageSpy.mockReset();
        getTeamSpy.mockReset();

    });

    it('should render with default props', async () => {
        getTeamSpy.mockImplementationOnce(() => Promise.resolve(mockState));

        let wrapper: RenderResult;
        act(() => {
            wrapper = renderComponent();
        });
        expect.assertions(1);

        await waitFor(() => {
            expect(wrapper.container).toMatchSnapshot();
        });
    });

    it('should get the team name from the TeamService and set the state', async () => {
        getTeamSpy.mockImplementationOnce(() => Promise.resolve(mockState));

        act(() => {
            renderComponent();
        });
        expect.assertions(2);

        await waitFor(() => {
            expect(getTeamSpy).toBeCalledTimes(1);
            expect(mockSetState).toBeCalledWith(mockState);
        });
    });

    it('should should display an error if the team cannot be retrieved', async () => {
        getTeamSpy.mockImplementationOnce(() => Promise.reject({
            response: {
                status: 500
            }
        }));

        act(() => {
            renderComponent();
        });

        await waitFor(() => {
            expect(getTeamSpy).toBeCalledTimes(1);
            expect(setErrorMessageSpy).toBeCalledWith({
                'description': 'There was an error retrieving the team.  Please try refreshing the page.',
                'title': 'Something went wrong'
            });
        });
    });
});
