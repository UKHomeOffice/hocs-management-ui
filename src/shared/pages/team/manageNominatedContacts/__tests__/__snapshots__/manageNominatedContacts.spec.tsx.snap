// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manage nominated contacts should render with default props 1`] = `
<div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-two-thirds-from-desktop"
    >
      <a
        class="govuk-back-link"
        href="/team-view/__teamId__"
      >
        Back
      </a>
      <h1
        class="govuk-heading-xl"
      >
        Manage nominated contacts
      </h1>
      <h2
        class="govuk-heading-l"
      >
        Team: __teamName__
      </h2>
    </div>
  </div>
  MOCK_ADD_NOMINATED_CONTACT_COMPONENT
  MOCK_CONTACT_LIST_COMPONENT
</div>
`;
